src/tools/remote-test-server/src/main.rs

```diff
@@ -277,9 +277,15 @@ fn handle_run(socket: TcpStream, work: &Path, tmp: &Path, lock: &Mutex<()>, conf
     let mut stderr = child.stderr.take().unwrap();
     let socket = Arc::new(Mutex::new(reader.into_inner()));
     let socket2 = socket.clone();
-    let thread = thread::spawn(move || my_copy(&mut stdout, 0, &*socket2));
-    my_copy(&mut stderr, 1, &*socket);
-    thread.join().unwrap();
+
+    if config.sequential {
+        my_copy(&mut stdout, 0, &*socket2);
+        my_copy(&mut stderr, 1, &*socket);
```

Comment:
This seems like it won't actually work? The reason we need threads is that my_copy uses blocking I/O -- this will fail (hang!) if the buffers fill up and the process expects to be able to continue writing...

Reply:
I have used this in https://github.com/rust-lang/rust/pull/100316, so I can say that it does work for UEFI. However, it might just be because of how io writing is implemented in UEFI since there are no threads.

Reply:
@Mark-Simulacrum So after looking around a bit more in the code, I'm not sure how my implementation is any worse than the current `sequential` implementation. If the sequential option is enabled, the `handle_run` function is launched in the main thread. This means that `my_copy` for stderr runs on the main thread anyway.

My commit just makes `my_copy` for stdout run on the `main` thread` as well. The only target I have tested this on is `UEFI` so maybe you can point to an example where this will lead to UB?

Reply:
This isn't going to lead to UB, it'll just hang -- and only if the process writes to both stdout and stderr in sufficient quantities to fill up the OS pipe.

Running one of the copies on the main thread is fine, but when we run both of them that leads to a problem, because only one of them is running at a time.

One approach to not use threads but solve this is to use a nonblocking API -- you can steal e.g. https://github.com/rust-lang/rust/blob/master/src/tools/compiletest/src/read2.rs to read/write from both on the same thread (always).